### PR TITLE
BiGCZ: Only search if query has changed

### PR DESCRIPTION
**Pull request by arottersman**
## Overview

We were searching any time you made a catalog tab active. Now we only run a search if we've updated the query (or the search has never been run).

While looking through data catalog views, noticed we were setting `className` dynamically, which I've had trouble with in the past. Correcting this fixed the bug on IE11 where the tab contents would remain the same for all tabs (ie all tabs showed CINERGI results).

Connects #1960 
Connects #2014  

### Demo

vujimkyvfj

## Testing Instructions

 * Pull and `bundle`
 * `localhost:800?bigcz`
On the major browsers:
  * Draw an AoI and search for something
  * Confirm the search doesn't execute again when you switch between tabs, but does re-execute if you update the query (either date filters or query text)
  * Confirm the pagination still works

### CODE REVIEW

**Comment by rajadain** [comment]
> immediately repeated promise

I think you mean "immediately resolved promise"

File: src/mmw/js/src/data_catalog/models.js [hunk 1]
```diff
@@ -29,6 +29,23 @@ var Catalog = Backbone.Model.extend({
         error: '',
     },
 
+    searchIfNeeded: function(query, fromDate, toDate, bounds) {
+        var self = this,
+            isSameSearch = query === this.get('query') &&
+                           fromDate === this.get('fromDate') &&
+                           toDate === this.get('toDate') &&
+                           utils.formatBounds(bounds) === this.get('bbox');
```
Comment (rajadain) [inline src/mmw/js/src/data_catalog/models.js]: Consider lifting `utils.formatBounds(bounds)` out of this and `search` below so it is called less frequently.

File: src/mmw/js/src/data_catalog/models.js [hunk 2]
```diff
@@ -29,6 +29,23 @@ var Catalog = Backbone.Model.extend({
...
+
+        if (!isSameSearch) {
```
Comment (rajadain) [inline src/mmw/js/src/data_catalog/models.js]: Consider the case when a search is already running. If a user were to change the query and re-search, `!isSameSearch` evaluates to `true` and we return `this.searchPromise`. However, when the first search returns, the `always` block executes and deletes `self.searchPromise`. What happens if this were to be deleted before the client can use its value?

Comment (rajadain) [inline src/mmw/js/src/data_catalog/models.js]: Ideally we would cancel `this.searchPromise` before starting a new one. Not sure how that can be done. Another alternative may be to chain the new search onto the old search.

If I'm misreading the code (which I could: promises are hard to follow), please do correct me.

**Comment by rajadain** [comment]
Should this connect #2014 instead of #2054? Seems to be making #2054 worse

File: src/mmw/js/src/data_catalog/models.js (hunk 2 above)
Comment (arottersman) [inline src/mmw/js/src/data_catalog/models.js]: I will work out a way to explicitly cancel the ongoing promise for clarity.

I don't think, however, that `delete self.searchPromise` prevents the new promise from resolving correctly (and we don't use `this.searchPromise` anywhere else). 

This is not what I'd expect the behavior to be, but here's a little test I made: 
```js
var self = this;
var d1 = $.Deferred();

this.myPromise = $.when(d1)
       .done(function (v) {
           console.log(v);
       })
       .always(function() {
	   console.log("deleting in first");
           delete self.myPromise;
       });

var d2 = $.Deferred();

this.myPromise = $.when(d2)
        .done(function (v) {
            console.log(v)
        })
        .always(function() {
	   console.log("deleting in second");
           delete self.myPromise;
        });

d1.resolve( "first done" );
d2.resolve( "second done" );

/** outputs
VM10771:6 first done
VM10771:9 deleting in first
VM10771:17 second done
VM10771:20 deleting in second
**/
```

**Comment by arottersman** [comment]
> Should this connect #2014 instead of #2054? Seems to be making #2054 worse

Yes, that must have been a typo. Thanks! (Although maybe #2054 will be fixed when this done?)

**Comment by arottersman** [comment]
This should be ready for another look! That double fixup corrects a mistake I made while rebasing.

**Comment by rajadain** [comment]
Taking another look

**Comment by arottersman** [comment]
Nice find — I've added a commit.

File: src/mmw/js/src/data_catalog/models.js [hunk 3]
```diff
@@ -29,10 +28,30 @@ var Catalog = Backbone.Model.extend({
...
+    searchIfNeeded: function(query, fromDate, toDate, bbox) {
+        var self = this,
+            isSameSearch = query === this.get('query') &&
+                           fromDate === this.get('fromDate') &&
+                           toDate === this.get('toDate') &&
+                           bbox === this.get('bbox');
+
+        if (!isSameSearch) {
+            if (this.searchPromise) {
+                this.searchPromise.reject({ superceded: true });
```
Comment (arottersman) [inline src/mmw/js/src/data_catalog/models.js]: Actually, this isn't going to work. `this.searchPromise` isn't a deferred object. It's a `jqXHR` because that's what backbone returns with `.fetch`. I think I'm going to change this to `abort`. Another option is to wrap the `fetch` in an deferred , but I think that makes it a little more difficult to follow.

**Comment by arottersman** [comment]
Added one more fixup to switch from `.reject()` to `.abort()` when cancelling the promise. Should be ready for another look now.

**Comment by arottersman** [comment]
Thanks for reviewing! Squashed fixups (and fixed typo in that commit message). Will merge once the build completes.
